ruby/src/api.cc

```diff
@@ -0,0 +1,529 @@
+        LOAD_SYMBOL(rb_block_call),
+        LOAD_SYMBOL(rb_funcall_passing_block),
+        LOAD_SYMBOL(rb_num2ulong),
+        LOAD_SYMBOL(rb_num2dbl),
+        LOAD_SYMBOL(rb_string_value_ptr),
+        LOAD_SYMBOL(rb_rescue2),
+        LOAD_SYMBOL(rb_protect),
+        LOAD_SYMBOL(rb_jump_tag),
+        LOAD_SYMBOL(rb_int2inum),
+        LOAD_SYMBOL(rb_enc_str_new),
+        LOAD_SYMBOL(rb_utf8_encoding),
+        LOAD_SYMBOL(rb_load),
+        LOAD_SYMBOL(rb_raise),
+        LOAD_SYMBOL(rb_block_proc),
+        LOAD_SYMBOL(rb_block_given_p),
+        LOAD_SYMBOL(rb_gc_register_address),
+        LOAD_SYMBOL(rb_gc_unregister_address),
+        LOAD_SYMBOL(rb_hash_foreach),
+        LOAD_SYMBOL(rb_define_attr),
+        LOAD_SYMBOL(rb_ivar_set),
+        LOAD_SYMBOL(rb_ivar_get),
+        LOAD_ALIASED_SYMBOL(rb_float_new_in_heap, rb_float_new),
+        LOAD_ALIASED_SYMBOL(rb_ary_new_capa, rb_ary_new2),
+        LOAD_SYMBOL(rb_ary_push),
+        LOAD_SYMBOL(rb_ary_entry),
+        LOAD_SYMBOL(rb_hash_new),
+        LOAD_SYMBOL(rb_hash_aset),
+        LOAD_SYMBOL(rb_hash_lookup),
+        LOAD_SYMBOL(rb_hash_lookup2),
+        LOAD_SYMBOL(rb_sym_to_s),
+        LOAD_SYMBOL(rb_to_id),
+        LOAD_SYMBOL(rb_id2name),
+        LOAD_SYMBOL(rb_define_alloc_func),
+        LOAD_SYMBOL(rb_data_object_alloc),
+        LOAD_SYMBOL(rb_gc_mark),
+        LOAD_SYMBOL(rb_yield_values),
+        LOAD_SYMBOL(rb_require),
+        LOAD_SYMBOL(rb_cObject),
+        LOAD_SYMBOL(rb_cArray),
+        LOAD_SYMBOL(rb_cHash),
+        LOAD_SYMBOL(rb_cString),
+        LOAD_SYMBOL(rb_cSymbol),
+        LOAD_SYMBOL(rb_cFixnum),
+        LOAD_SYMBOL(rb_cFloat),
+        LOAD_SYMBOL(rb_eException),
+        LOAD_SYMBOL(rb_eArgError),
+        LOAD_SYMBOL(rb_eTypeError),
+        LOAD_SYMBOL(rb_eStandardError),
+        LOAD_SYMBOL(rb_eRuntimeError),
+        LOAD_OPTIONAL_SYMBOL(ruby_setup),
+        LOAD_SYMBOL(ruby_init),
+        LOAD_SYMBOL(ruby_options),
+        LOAD_SYMBOL(ruby_cleanup),
+        _library(move(library))
+    {
+    }
+
+    api::~api()
+    {
+        // API is shutting down; free all remaining data objects
+        // Destructors may unregister the data object, so increment the iterator before freeing
+        for (auto it = _data_objects.begin(); it != _data_objects.end();) {
+            auto data = reinterpret_cast<RData*>(*it);
+            ++it;
+            if (data->dfree) {
+                data->dfree(data->data);
+                data->dfree = nullptr;
+            }
+        }
+        if (_initialized && _library.first_load() && cleanup) {
+            ruby_cleanup(0);
+        }
+    }
+
+    api* api::instance()
+    {
+        static unique_ptr<api> instance = create();
+        return instance.get();
+    }
+
+    unique_ptr<api> api::create()
+    {
+        lth_lib::dynamic_library library = find_library();
+        if (!library.loaded()) {
+            LOG_WARNING("could not locate a ruby library");
```

Comment:
We still need a method for handling info/debug messages (LTH-29), but for the warnings I think we should instead throw exceptions.